### DISCUSSION TITLE
Raise minimum support for clang to 3.9 (early)?

### DISCUSSION THREAD

**Discussion by fabratu** [networkit/networkit#703] [General]
Hi,

NetworKit aims to support compiler-version, which were released less than five years ago. However we have some pull requests in the pipeline, which need at least clang 3.9 due to bugs in clang 3.8 and below.

Either we wait until September until the waiting time is over or we raise the minimum support early. From the outside looking in, it seems that clang 3.8 is not really used anymore. At least a check in the linux/unix eco-system showed, that only a small margin of OS-distributions (which are not EOL) use this or an older version.

Ubuntu 16.04 for example shipped originally with 3.8, but has since updated to 6.0. One major dist still providing an old 3.4 clang version is vanilla CentOS 7 (without SCL or EPEL). Another minor dist would be Debian 8 (for commercial ELTS), which ships with 3.5, and Ubuntu 14.04, which ships with 3.8 in recent updates. It is likely however, that most professional or semi-professional environments use additional package repos or features like modules in order to provide more modern versions.

If there are reasons / use cases to keep clang 3.8 let us know.

Thanks!

**Comment by avdgrinten** [comment]
I think Debian 8 is not relevant anymore; that's two releases behind the current stable one. CentOS 7 might still be relevant in HPC environments. Does CentOS 7 have official repositories that contain newer Clang versions?

**Reply by fabratu** [reply]
EPEL (approved third-party) has 3.9 and SCL (not by default enabled CentOS repo) has 7.

EPEL = Extra Packages for Enterprise Linux
SCL = Software Collections

**Reply by avdgrinten** [reply]
Does the same hold for RHEL 7?

I think if there are widely-used CentOS/RHEL 7 repos that contain a Clang > 3.8, upgrading Clang is reasonable.

What's the maximal Clang version that is widely available on all current distros (younger than 5 years)? Clang 3.9?

**Reply by fabratu** [reply]
Yes, RHEL 7 uses the same repos + more commercial ones (don't know the specifics though).

For the max. clang version: If we exclude EOL and Arch-distros it boils down to LTS-versions and flavors. Other distros are much more recent and have like clang 7, 8 or newer as default. For Debian 9 and Ubuntu 16.04 it's clang 4, NetBSD 8.2 has 9. One unknown player is SLES which does not have a public repo (search) available. From their release notes it seems, that the regularly update the underlying libraries and packages. Don't know if this includes llvm/clang.

**Comment by hmeyerhenke** [comment]
I think it would be okay this time (because there is a good reason) to not wait 5 years. You could ask on the mailing list if someone has a problem with that (no answer within x days = okay). I'd be surprised if someone said in this particular case that a longer support is necessary. In general the 5 years time span is good, though.
